rules/TYPO312/v4/MigrateTypoScriptFrontendControllerTypeRector.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Ssch\TYPO3Rector\TYPO312\v4;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\PropertyFetch;
+use Rector\Rector\AbstractRector;
+use Ssch\TYPO3Rector\NodeFactory\Typo3GlobalsFactory;
+use Ssch\TYPO3Rector\NodeResolver\Typo3NodeResolver;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @changelog https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.4/Deprecation-100405-PropertyTypoScriptFrontendController-type.html
+ * @see \Ssch\TYPO3Rector\Tests\Rector\v12\v4\MigrateTypoScriptFrontendControllerTypeRector\MigrateTypoScriptFrontendControllerTypeRectorTest
+ */
+final class MigrateTypoScriptFrontendControllerTypeRector extends AbstractRector
+{
+    /**
+     * @var string
+     */
+    private const TYPE = 'type';
+
+    /**
+     * @readonly
+     */
+    private Typo3NodeResolver $typo3NodeResolver;
+
+    /**
+     * @readonly
+     */
+    private Typo3GlobalsFactory $typo3GlobalsFactory;
+
+    public function __construct(
+        Typo3NodeResolver $typo3NodeResolver,
+        Typo3GlobalsFactory $typo3GlobalsFactory
+    ) {
+        $this->typo3NodeResolver = $typo3NodeResolver;
+        $this->typo3GlobalsFactory = $typo3GlobalsFactory;
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Migrate TypoScriptFrontendController->type', [new CodeSample(
+            <<<'CODE_SAMPLE'
+$GLOBALS['TSFE']->type;
+CODE_SAMPLE
+            ,
+            <<<'CODE_SAMPLE'
+$GLOBALS['TSFE']->getPageArguments()->getPageType();
+CODE_SAMPLE
+        )]);
+    }
+
+    public function getNodeTypes(): array
+    {
+        return [PropertyFetch::class];
+    }
+
+    /**
+     * @param PropertyFetch $node
+     * @return Node|null
+     */
+    public function refactor(Node $node)
+    {
+        if ($this->shouldSkip($node)) {
+            return null;
+        }
+
+        if (! $this->isNames($node->name, [self::TYPE])) {
+            return null;
+        }
+
+        return $this->createGetPageTypeMethodCall(
+            $this->createTSFEGetPageArgumentsMethodCall()
+        );
+    }
+
+    private function shouldSkip(PropertyFetch $propertyFetch): bool
```

Comment:
This is not the only case. You could also have cases where some is using a wrapper method. You should also check for the method call on TypoScriptFrontendController

```php
private function getTypoScriptFrontendController(): TypoScriptFrontendController
{
    return $GLOBALS['TSFE'];
}